ArduCopter/tuning.cpp

```diff
@@ -1,215 +1,201 @@
+const uint8_t AP_Tuning_Copter::tuning_set_rate_pitch[] =      { TUNING_RATE_PITCH_KD, TUNING_RATE_PITCH_KPI };
+const uint8_t AP_Tuning_Copter::tuning_set_rate_yaw[] =        { TUNING_RATE_YAW_KP, TUNING_RATE_YAW_KI, TUNING_RATE_YAW_KD };
+const uint8_t AP_Tuning_Copter::tuning_set_ang_roll_pitch[] =  { TUNING_ANG_ROLL_KP, TUNING_ANG_PITCH_KP };
+const uint8_t AP_Tuning_Copter::tuning_set_ang_yaw[] =         { TUNING_ANG_YAW_KP };
+const uint8_t AP_Tuning_Copter::tuning_set_pxy[] =             { TUNING_PXY_P };
+const uint8_t AP_Tuning_Copter::tuning_set_pz[] =              { TUNING_PZ_P };
+const uint8_t AP_Tuning_Copter::tuning_set_vxy[] =             { TUNING_VXY_P, TUNING_VXY_I };
+const uint8_t AP_Tuning_Copter::tuning_set_az[] =              { TUNING_AZ_P, TUNING_AZ_I, TUNING_AZ_D };
+
+// macro to prevent getting the array length wrong
+#define TUNING_ARRAY(v) ARRAY_SIZE(v), v
+
+// list of tuning sets
+const AP_Tuning_Copter::tuning_set AP_Tuning_Copter::tuning_sets[] = {
+    { TUNING_SET_RATE_ROLL_PITCH, TUNING_ARRAY(tuning_set_rate_roll_pitch) },
+    { TUNING_SET_RATE_ROLL,       TUNING_ARRAY(tuning_set_rate_roll) },
+    { TUNING_SET_RATE_PITCH,      TUNING_ARRAY(tuning_set_rate_pitch) },
+    { TUNING_SET_RATE_YAW,        TUNING_ARRAY(tuning_set_rate_yaw) },
+    { TUNING_SET_ANG_ROLL_PITCH,  TUNING_ARRAY(tuning_set_ang_roll_pitch) },
+    { TUNING_SET_ANG_YAW,         TUNING_ARRAY(tuning_set_ang_yaw) },
+    { TUNING_SET_PXY,             TUNING_ARRAY(tuning_set_pxy) },
+    { TUNING_SET_PZ,              TUNING_ARRAY(tuning_set_pz) },
+    { TUNING_SET_VXY,             TUNING_ARRAY(tuning_set_vxy) },
+    { TUNING_SET_AZ,              TUNING_ARRAY(tuning_set_az) },
+    { TUNING_SET_NONE, 0, nullptr }
+};
 
-// tuning - updates parameters based on the ch6 tuning knob's position
-//  should be called at 3.3hz
-void Copter::tuning() {
-
-    // exit immediately if not using tuning function, or when radio failsafe is invoked, so tuning values are not set to zero
-    if ((g.radio_tuning <= 0) || failsafe.radio || failsafe.radio_counter != 0 || g.rc_6.radio_in == 0) {
-        return;
-    }
-
-    // set tuning range and then get new value
-    g.rc_6.set_range_in(g.radio_tuning_low,g.radio_tuning_high);
-    float tuning_value = (float)g.rc_6.control_in / 1000.0f;
-    // Tuning Value should never be outside the bounds of the specified low and high value
-    tuning_value = constrain_float(tuning_value, g.radio_tuning_low/1000.0f, g.radio_tuning_high/1000.0f);
-
-    Log_Write_Parameter_Tuning(g.radio_tuning, tuning_value, g.rc_6.control_in, g.radio_tuning_low, g.radio_tuning_high);
+/*
+  table of tuning names
+ */
+const AP_Tuning_Copter::tuning_name AP_Tuning_Copter::tuning_names[] = {
+    { TUNING_RATE_ROLL_KPI, "RateRollPI" },
+    { TUNING_RATE_ROLL_KD,  "RateRollD" },
+    { TUNING_RATE_PITCH_KPI,"RatePitchPI" },
+    { TUNING_RATE_PITCH_KD, "RatePitchD" },
+    { TUNING_NONE, nullptr }
```

Comment:
yes, I'm adding those now, along with the ability to tune any single parameter without a selector switch